monash-human-power/dashboard
# Allow uploading of config bundle

## Description

In this second part of making the `BOOST` page functional, the PR brings to you the ability to upload a config 'bundle' (I kept the name same as on notion) which is essentially a `JSON` containing the 4 configs (`powerPlan, rider, bike, track`). The bundle is separated into the 4 configs and then sent over `MQTT` on `boost/configs/action`. 

I expected it to be a very small PR, but ended up making it a little bigger thanks to adding a bunch of Error handing code. There's a lot of `if` statements to achieve this, feel free to suggest any better ways to do it. I kind of feel I went overboard with checking edge cases, but I did it so we don't have to ever worry about it later (especially if we keep the individual upload buttons and the 'Upload all configs' button, there are chances users will use them the wrong way around).

This PR adds Error handling for the following:
- When a bundle does not contain each of `powerPlan, rider, bike and track`
- When a bundle does not contain exactly 4 configs
- When a config either uploaded separately or inside a bundle has a name which is already used and displayed 

I went with using `toasts` for simple success messages (thanks to @Blake-Haydon for suggesting the sick-as toast library). And for error messages, I used `Modals`. TBH react `Modal` is quite ugly, but didn't spend too much time trying to find alternatives. On that note, I'm aware that the error messages don't look that great inside the `Modal` and so I was thinking about using JSX Elements to display it nicely using bullet points but it was taking quite some time. Let me know if you prefer for the message to be nicely formatted and I can work on it.

I have also add 3 test bundle files which you can easily upload and check that everything is working without having to change the bundle content to force edge cases. I'll remove them perhaps when the entire `BOOST` page is functional and we'll no longer need to keep checking these edge cases.

## Screenshots

When uploading a bundle containing repeated config names:
<img width="915" alt="Screen Shot 2021-03-03 at 11 46 43 am" src="https://user-images.githubusercontent.com/63642262/109888915-7ef99c80-7cd8-11eb-8dc2-3983529232da.png">

Upon uploading a correct bundle:
<img width="1275" alt="Screen Shot 2021-03-04 at 10 04 45 am" src="https://user-images.githubusercontent.com/63642262/109888941-87ea6e00-7cd8-11eb-9c29-701eebd71b0e.png">


## Steps to Test

1) In `client/` run `yarn start`
2) In the `BOOST` page try uploading the 3 differentiations scenario bundles (provided in `client`) and check that each is dealt with correctly
3) Upload an individual config as well and see that nothing form the last PR is broken
4) Might as well run the server (inside `server` run `yarn start`) and the `MQTT` subscription `boost/configs/action` and check that things are being sent correctly.
